src/components/MarqueeSelection/MarqueeSelection.tsx

```diff
@@ -0,0 +1,252 @@
+import * as React from 'react';
+import { AutoScroll } from '../../utilities/AutoScroll/AutoScroll';
+import { BaseComponent } from '../../common/BaseComponent';
+import { assign } from '../../utilities/object';
+import { css } from '../../utilities/css';
+import { findScrollableParent } from '../../utilities/scrollUtilities';
+import {
+  IMarqueeSelectionProps
+} from './MarqueeSelection.Props';
+import './MarqueeSelection.scss';
+
+export interface IPoint {
+  x: number;
+  y: number;
+}
+
+export interface IRectangle {
+  left: number;
+  top: number;
+  width: number;
+  height: number;
+
+  right?: number;
+  bottom?: number;
+}
+
+export interface IMarqueeSelectionState {
+  dragOrigin?: IPoint;
+  dragRect?: IRectangle;
+}
+
+// We want to make the marquee selection start when the user drags a minimum distance. Otherwise we'd start
+// the drag even if they just click an item without moving.
+const MIN_DRAG_DISTANCE = 5;
+
+/**
+ * MarqueeSelection component abstracts managing a draggable rectangle which sets items selected/not selected.
+ * Elements which have data-selectable-index attributes are queried and measured once to determine if they
+ * fall within the bounds of the rectangle. The measure is memoized during the drag as a performance optimization
+ * so if the items change sizes while dragging, that could cause incorrect results.
+ */
+export class MarqueeSelection extends BaseComponent<IMarqueeSelectionProps, IMarqueeSelectionState> {
+  public static defaultProps = {
+    rootTagName: 'div',
+    rootProps: {}
+  };
+
+  public refs: {
+    [key: string]: React.ReactInstance;
+    root: HTMLElement;
+  };
+
+  private _dragOrigin: IPoint;
+  private _rootRect: IRectangle;
+  private _lastMouseEvent: MouseEvent;
+  private _autoScroll;
+  private _selectedIndicies: { [key: string]: boolean };
+  private _itemRectCache: { [key: string]: IRectangle };
+
+  constructor(props: IMarqueeSelectionProps) {
+    super(props);
+
+    this.state = {
+      dragRect: null
```

Comment:
I strongly suggest using `undefined` instead of `null` for uninitialized values.